designs/2022-supress-ignored-file-warnings/README.md

```diff
@@ -0,0 +1,174 @@
+
+For example, with
+
+```
+// .eslintignore
+foo.js
+```
+
+Running
+
+```
+eslint foo.js
+```
+
+Will result in
+
+```
+warning  File ignored because of a matching ignore pattern. Use "--no-ignore" to override
+```
+
+It is possible to filter out ignored files using ESLint `CLIEngine`. This was the main reasoning to reject the CLI flag idea during [TSC meeting in 2018](https://gitter.im/eslint/tsc-meetings/archives/2018/08/02). The commitee was only considering the use case of integrations. However, this problem affects end users as well. Lint-staged FAQ includes a section [how can I ignore files from .eslintignore](https://github.com/okonet/lint-staged#how-can-i-ignore-files-from-eslintignore), which suggests adding additional config file and using `CLIEngine.isPathIgnored` to filter out ignored files. Having a CLI flag for this would improve the end-user experience.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+ESLint CLI should have an optional flag to disable the ignored file warning. The flag should be boolean. In order to stay consistent with current CLI flags, it should be called `--no-warning-on-ignored-files`. If it is present, ESLint should supress the `File ignored because of a matching ignore pattern.` warning. The flag should replicate `eslint.lintText`[warnIgnored option](https://eslint.org/docs/developer-guide/nodejs-api#-eslintlinttextcode-options) when it is set to `false`.
+
+The current warning should be updated to `File ignored because of a matching ignore pattern. Use "--no-ignore" to override. Use "--no-warning-on-ignored-files" to suppress this warning.`
+
+Add the flag to `optionator()` in `options.js`:
+
+```js
+{
+    option: "warning-on-ignored-files",
+    type: "Boolean",
+    default: "true",
+    description: "Prevent warnings when file list includes ignored files"
+},
+```
+
+It should be placed either unde `heading: "Miscellaneous"`, or `heading: "Handling warnings"`, or `heading: "Ignoring files"`.
+
+Add `warningOnIgnoredFiles` to `translateOptions()` in `cli.js`. It should be returned as is.
+
+Destructure `warningOnIgnoredFiles` from `options` from `internalSlotsMap.get(this)` in `executeOnFiles()` function in `cli-engine.js`.
+
+In `iterateFiles()` of `executeOnFiles()` function in `cli-engine.js`, add additional condition `if (ignored && warningOnIgnoredFiles)` to prevent warnings for ignored files.
```

Comment:
Will `executeOnText()` ignore the new constructor option and only use its parameter `warnIgnored` when determining whether the warning should be reported?

Reply:
In my opinion – it would make sense to deprecate `executeOnText`'s `warnIgnored` and replace it with `warningOnIgnoredFiles` constructor option. However, that would be a breaking change. Would like to get some more opinions on this

Reply:
It seems like the warnIgnored parameter should default to the value of warningOnIgnoredFiles?

Reply:
That's a good middle ground

Reply:
They have different defaults: the parameter has `false`, while the new constructor option will have `true` as the default value. If we change the parameter to default to the value of the constructor option (which defaults to `true`), it's still a breaking change.

Reply:
As I was working on the PR, I noticed some inconsistencies with the defaults.
- `cli.js` forces `warnIgnored: true` when text is passed via stdin [here](https://github.com/eslint/eslint/blob/f31216a90a6204ed1fd56547772376a10f5d3ebb/lib/cli.js#L298)
- `eslint.js` defaults to `warnIgnored = false` in `lintText` [here](https://github.com/eslint/eslint/blob/main/lib/eslint/eslint.js#L571)
- `eslint.js` current implicitly shows the warning on `lintFiles` (the whole point of this RFC is to make that configurable), therefore it can be considered it defaults to `warnIgnored = true`. The same applies to `cli-engine.js` `executeOnFiles`, it implicitly defaults to `warnIgnored = true`.
- `cli-engine.js` does not have a default `warnIgnored` value in `executeOnText`, so when that parameter is omitted it is a falsy `undefined`. So we could say it defaults to `warnIgnored = false`

I propose to make this consistent by defaulting to `warnIgnored: true` in `default-cli-options.js` and removing other fallbacks/overrides. That way all cases above would consistently default to `true`.

I suppose that's a breaking change, though.

Reply:
New design only affects `lintFiles` to maintain backwards compatibility